src/System.IO.Compression/src/System/IO/Compression/DeflateManaged/InflaterManaged.cs

```diff
@@ -134,10 +137,26 @@ public int Inflate(byte[] bytes, int offset, int length)
                     _formatReader.Validate();
                 }
             }
+            count = RestrictStreamWithUnCompressedSize(count, out bool wasOverLimit);
+
+            if (wasOverLimit)
```

Comment:
I think that this variable can be eliminated...

Based on the fact that prior blocks could have already been processed, it seems like reading up to the limit should work.  So `count = RestrictStreamWithUnCompressedSize(count)`.  If that returns 0 (meaning that the total number of bytes declared in the metadata had already been returned) then throw the exception.

Reply:
`count = 0` doesn't mean it was over limit nor it is invalid read, in the DeflateStream.Read() it will be 0 as in the firs round of the loop the inflator input will be empty => nothing available in iflater output https://github.com/dotnet/corefx/blob/837cc58bb33f545c58375955909dee27f07d5d3d/src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs#L268-L298

Reply:
OK, but the situation I'm looking to avoid is that sometimes you get the header value, and sometimes you get less... because the current implementation seems like it will throw once it hits the limit, not on the first read after.

If that's a situation that's impossible to avoid, then perhaps a better flow would be

* If this read will exceed the limit
  * Copy bytes up to the limit
  * Set a field that says this inner stream is now invalid
  * The next call into this object throws the InvalidDataException.

By analogy, this is "walk out peacefully, but lock the door behind you".

Reply:
Yeah I was actually thinking to use Finished flag instead of throwing